src/server/frontend_xwayland/xwayland_server.cpp

```diff
@@ -0,0 +1,296 @@
+    addr.sun_family = AF_LOCAL;
+    socket_fd = socket(PF_LOCAL, SOCK_STREAM | SOCK_CLOEXEC, 0);
+
+    name_size = snprintf(addr.sun_path, sizeof addr.sun_path, "/tmp/.X11-unix/X%d", xdisplay) + 1;
+    size = offsetof(struct sockaddr_un, sun_path) + name_size;
+    unlink(addr.sun_path);
+
+    if (bind(socket_fd, (struct sockaddr *)&addr, size) < 0)
+    {
+        mir::log_error("Failed to bind to x11 socket");
+        close(socket_fd);
+        return;
+    }
+
+    if (listen(socket_fd, 1) < 0)
+    {
+        mir::log_error("Failed to listen to x11 socket");
+        unlink(addr.sun_path);
+        close(socket_fd);
+        return;
+    }
+}
+
+void mf::XWaylandServer::bind_to_abstract_socket()
+{
+    struct sockaddr_un addr;
+    socklen_t size, name_size;
+    addr.sun_family = AF_LOCAL;
+
+    abstract_socket_fd = socket(PF_LOCAL, SOCK_STREAM | SOCK_CLOEXEC, 0);
+
+    name_size = snprintf(addr.sun_path, sizeof addr.sun_path, "%c/tmp/.X11-unix/X%d", 0, xdisplay);
+    size = offsetof(struct sockaddr_un, sun_path) + name_size;
+    if (bind(abstract_socket_fd, (struct sockaddr *)&addr, size) < 0)
+    {
+        mir::log_error("Failed to bind to abstract socket");
+        close(abstract_socket_fd);
+        return;
+    }
+
+    if (listen(abstract_socket_fd, 1) < 0)
+    {
+        mir::log_error("Failed to listen to abstract socket");
+        close(abstract_socket_fd);
+        return;
+    }
+}
+
+// TODO this can be written with more modern c++
+int mf::XWaylandServer::create_lockfile()
+{
+    char lockfile[256];
+    char pid[11];
+    int fd, size;
+    pid_t other;
+
+    snprintf(lockfile, sizeof lockfile, "/tmp/.X%d-lock", xdisplay);
+    fd = open(lockfile, O_WRONLY | O_CLOEXEC | O_CREAT | O_EXCL, 0444);
+    if (fd < 0 && errno == EEXIST)
+    {
+        fd = open(lockfile, O_CLOEXEC | O_RDONLY);
+        if (fd < 0 || read(fd, pid, 11) != 11)
+        {
+            if (fd >= 0)
+                close(fd);
+
+            errno = EEXIST;
+            return -1;
+        }
+
+        /* Trim the trailing LF, or at least ensure it's NULL. */
+        pid[10] = '\0';
+
+        other = std::stoi(pid);
+
+        if (kill(other, 0) < 0 && errno == ESRCH)
+        {
+            /* stale lock file; unlink and try again */
+            close(fd);
+            if (unlink(lockfile))
+                /* If we fail to unlink, return EEXIST
+                   so we try the next display number.*/
+                errno = EEXIST;
+            else
+                errno = EAGAIN;
+            return -1;
+        }
+
+        close(fd);
+        errno = EEXIST;
+        return -1;
+    }
+    else if (fd < 0)
+    {
+        return -1;
+    }
+
+    size = dprintf(fd, "%10d\n", getpid());
+    if (size != 11)
+    {
+        unlink(lockfile);
+        close(fd);
+        return -1;
+    }
+
+    close(fd);
+
+    return 0;
+}
+
+void mf::XWaylandServer::setup_socket()
+{
+// TODO I dont really like goto's, should remove
```

Comment:
I really don't like ignoring post-condition failures by logging an error and returning, and that design choice is why you end up with a goto here.